drivers/at_cmd/at_cmd.c

```diff
@@ -51,6 +59,93 @@ struct callback_work_item {
 K_MEM_SLAB_DEFINE(rsp_work_items, sizeof(struct callback_work_item),
 		  CONFIG_AT_CMD_RESPONSE_BUFFER_COUNT, 4);
 
+/**@brief Find notification handler.
+ *
+ * @note  Returns next and writes previous to prev.
+ */
+static notif_handler_t *find_notif_handler(sys_slist_t *list,
+	at_cmd_handler_t handler, notif_handler_t **prev_out)
+{
+	notif_handler_t *prev = NULL;
+	notif_handler_t *curr = NULL;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(list, curr, node) {
+		if (curr->handler == handler) {
+			*prev_out = prev;
+			return curr;
+		}
+		prev = curr;
+	}
+	return NULL;
+}
+
+/**@brief Append notification handler. */
+static int append_notif_handler(sys_slist_t *list, at_cmd_handler_t handler)
+{
+	notif_handler_t *to_ins;
+
+	k_mutex_lock(&list_mtx, K_FOREVER);
+
+	/* Check if it exists */
+	if (find_notif_handler(list, handler, &to_ins) != NULL) {
+		LOG_DBG("Already in the list.\n");
+		k_mutex_unlock(&list_mtx);
+		return 0;
+	}
+
+	/* Allocate memory and fill. */
+	to_ins = (notif_handler_t *)k_malloc(sizeof(notif_handler_t));
+	if (to_ins == NULL) {
+		k_mutex_unlock(&list_mtx);
+		return -ENOBUFS;
+	}
+	memset(to_ins, 0, sizeof(notif_handler_t));
+	to_ins->handler = handler;
+
+	/* Append. */
+	sys_slist_append(list, &to_ins->node);
+
+	k_mutex_unlock(&list_mtx);
+	return 0;
+}
+
+/**@brief Remove notification handler. */
+static void remove_notif_handler(sys_slist_t *list, at_cmd_handler_t handler)
+{
+	notif_handler_t *curr;
+	notif_handler_t *prev = NULL;
+
+	k_mutex_lock(&list_mtx, K_FOREVER);
+
+	/* Check if it exists */
+	curr = find_notif_handler(list, handler, &prev);
+	if (curr == NULL) {
+		LOG_DBG("Not found.\n");
+		k_mutex_unlock(&list_mtx);
+		return;
+	}
+
+	/* Remove */
+	sys_slist_remove(list, &prev->node, &curr->node);
+	k_free(curr);
+
+	k_mutex_unlock(&list_mtx);
+}
+
+/**@brief Dispatcher for notification handlers. */
+static void notification_handler (char *response)
+{
+	notif_handler_t *curr;
+
+	k_mutex_lock(&list_mtx, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&notif_handler_list, curr, node) {
```

Comment:
We've agreed to use `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` here.

Reply:
My bad. I will fix in the next upload.